src/mito2/src/compaction/memory_manager.rs

```diff
@@ -0,0 +1,285 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::fmt;
+use std::sync::Arc;
+use std::sync::atomic::{AtomicU64, Ordering};
+
+use serde::{Deserialize, Serialize};
+use tokio::sync::{Notify, Semaphore, TryAcquireError};
+
+use crate::metrics::{
+    COMPACTION_MEMORY_IN_USE, COMPACTION_MEMORY_LIMIT, COMPACTION_MEMORY_REJECTED,
+};
+
+/// Minimum bytes controlled by one semaphore permit.
+const PERMIT_GRANULARITY_BYTES: u64 = 1 << 20; // 1 MB
+
+/// Defines how the scheduler reacts when compaction cannot acquire enough memory.
+#[derive(Debug, Default, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
+#[serde(rename_all = "snake_case")]
+pub enum OnExhaustedPolicy {
+    /// Wait until enough memory is released.
+    #[default]
+    Wait,
+    /// Skip the compaction silently.
+    Skip,
+    /// Fail the compaction request immediately.
+    Fail,
+}
+
+/// Global memory manager for compaction tasks.
+#[derive(Clone)]
+pub struct CompactionMemoryManager {
+    inner: Option<Arc<Inner>>,
+}
+
+struct Inner {
+    semaphore: Semaphore,
+    limit_bytes: u64,
+    used_bytes: AtomicU64,
+    notify: Notify,
+}
+
+impl CompactionMemoryManager {
+    /// Creates a new memory manager with the given limit in bytes.
+    /// `limit_bytes = 0` disables the limit.
+    pub fn new(limit_bytes: u64) -> Self {
+        if limit_bytes == 0 {
+            COMPACTION_MEMORY_LIMIT.set(0);
+            return Self { inner: None };
+        }
+
+        let permits = bytes_to_permits(limit_bytes) as usize;
+        COMPACTION_MEMORY_LIMIT.set(limit_bytes as i64);
+
+        Self {
+            inner: Some(Arc::new(Inner {
+                semaphore: Semaphore::new(permits),
+                limit_bytes,
+                used_bytes: AtomicU64::new(0),
+                notify: Notify::new(),
+            })),
+        }
+    }
+
+    /// Returns the configured limit in bytes (0 if unlimited).
+    pub fn limit_bytes(&self) -> u64 {
+        self.inner
+            .as_ref()
+            .map(|inner| inner.limit_bytes)
+            .unwrap_or(0)
+    }
+
+    /// Returns currently used bytes.
+    pub fn used_bytes(&self) -> u64 {
+        self.inner
+            .as_ref()
+            .map(|inner| inner.used_bytes.load(Ordering::Relaxed))
+            .unwrap_or(0)
+    }
+
+    /// Returns available bytes.
+    pub fn available_bytes(&self) -> u64 {
+        self.limit_bytes().saturating_sub(self.used_bytes())
+    }
```

Comment:
Since we can get available permits from the semaphore, can we just use the aligned bytes as the semaphore's max_permits? We only need to align the bytes before acquiring. Then the available_permits always indicate the available memory.
https://docs.rs/tokio/latest/tokio/sync/struct.Semaphore.html#method.available_permits